### ISSUE TITLE
UnsupportedMessageTypeException: io.netty.handler.codec.http.DefaultFullHttpRequest (expected: io.netty.buffer.ByteBuf)

### ISSUE THREAD

**Issue by asarkar** [reactor/reactor-netty#229]
### Expected behavior
It should work, right?
### Actual behavior
```
io.netty.handler.codec.UnsupportedMessageTypeException: io.netty.handler.codec.http.DefaultFullHttpRequest (expected: io.netty.buffer.ByteBuf)
	at io.netty.handler.ssl.SslHandler.write(SslHandler.java:710) ~[netty-handler-4.1.17.Final.jar!/:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738) ~[netty-transport-4.1.17.Final.jar!/:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:801) ~[netty-transport-4.1.17.Final.jar!/:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:814) ~[netty-transport-4.1.17.Final.jar!/:4.1.17.Final]
```
### Steps to reproduce
I'm trying to set up `WebClient` to use a certificate on the client side, equivalent to `curl --cacert` option. This is turning out to be surprisingly difficult.
```
try (InputStream is = new FileInputStream(properties.getCertFile())) {
    CertificateFactory cf = CertificateFactory.getInstance("X.509");
    X509Certificate cacert = (X509Certificate) cf.generateCertificate(is);

    tmf = TrustManagerFactory
            .getInstance(TrustManagerFactory.getDefaultAlgorithm());
    KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
    ks.load(null); // we don't need the KeyStore instance to come from a file.
    ks.setCertificateEntry("cacert", cacert);

    tmf.init(ks);
}

SslContext sslContext = SslContextBuilder.forClient()
        .trustManager(tmf)
        .build();

ClientHttpConnector httpConnector = new ReactorClientHttpConnector(opt -> opt.afterChannelInit(channel -> {
    ChannelPipeline pipeline = channel.pipeline();
    SslHandler sslHandler = sslContext.newHandler(channel.alloc());
    SSLEngine sslEngine = sslHandler.engine();
    SSLParameters sslParameters = sslEngine.getSSLParameters();
    sslParameters.setEndpointIdentificationAlgorithm("HTTPS");
    sslEngine.setSSLParameters(sslParameters);

    pipeline.addLast("ssl", sslHandler);
}));
builder = WebClient.builder()
        .clientConnector(httpConnector);
```
### Reactor Netty version
4.1.17.Final, Spring Boot 2.0.0.BUILD-SNAPSHOT
### JVM version (e.g. `java -version`)
1.8.0_66
### OS version (e.g. `uname -a`)
macOS 10.13.1

**Comment by violetagg** [comment]
Hi,

You might want to add the SslHandler as first in the pipeline [1].
Can you try to change this `pipeline.addLast("ssl", sslHandler);` to this `pipeline.addFirst("ssl", sslHandler);`?

Regards,
Violeta

[1] https://github.com/netty/netty/blob/4.1/example/src/main/java/io/netty/example/securechat/SecureChatClientInitializer.java#L39

**Comment by asarkar** [comment]
@violetagg Tried that too, then I get another exception, "not SSL" (not exact). Do I need the other handlers shown in the example too? Do you've any tests/examples that do what I'm trying to do, which is, set up a client cert?

**Comment by violetagg** [comment]
No you do not need the handlers from the example, it was just a pointer that you need SslHandler to be added as a first handler in the pipeline.

Try this example and let us know the result:
- https://github.com/reactor/reactor-netty/blob/master/src/test/java/reactor/ipc/netty/http/client/HttpClientTest.java#L726

**Comment by asarkar** [comment]
Updated to
```
ClientHttpConnector httpConnector = new ReactorClientHttpConnector(opt -> 
    opt.afterNettyContextInit(ctx -> {
        opt.connectAddress(() -> ctx.address())
                .sslContext(sslContext);
}));
```

And now another error:

```
javax.net.ssl.SSLException: SSLEngine closed already
	at io.netty.handler.ssl.SslHandler.wrap(...)(Unknown Source) ~[netty-handler-4.1.17.Final.jar!/:4.1.17.Final]
```

**Comment by violetagg** [comment]
Can you try to use directly Reactor Netty instead of WebClient or you depend on Spring Framework?

**Comment by asarkar** [comment]
The project is Spring based, using Netty directly is out of question.

**Comment by violetagg** [comment]
> The project is Spring based, using Netty directly is out of question.
ok

```
ClientHttpConnector httpConnector = new ReactorClientHttpConnector(opt -> 
    opt.afterNettyContextInit(ctx -> {
        opt.connectAddress(() -> ctx.address())
                .sslContext(sslContext);
}));
```
this is not OK

Try the code below with enabled log level `reactor.ipc.netty=debug` so that we can investigate the problem

`new ReactorClientHttpConnector(opt -> opt.sslContext(ctx))`
or
`new ReactorClientHttpConnector(opt -> opt.connectAddress(() -> ctx.address())
                .sslContext(sslContext))`

if you can provide some simple example it will be great

**Comment by asarkar** [comment]
@violetagg The code snippets you shown are not valid - there is no `ctx` variable in scope

**Comment by violetagg** [comment]
Which does not work this:
`new ReactorClientHttpConnector(opt -> opt.sslContext(sslContext))`
or
`new ReactorClientHttpConnector(opt -> opt.connectAddress(() -> ctx.address()) .sslContext(sslContext))` 

this `ctx.address())` is the server address

**Comment by asarkar** [comment]
OK, I thought in `opt.sslContext(ctx)`, `ctx` is the `NettyContext`. In the 2nd snippet, `ctx` is infact the `NettyContext` and it is not available. I'd have to call `afterNettyContextInit` as I showed [earlier](https://github.com/reactor/reactor-netty/issues/229#issuecomment-347836654).

That said, enabling debug logs helped me find the problem, and I feel incredibly stupid. I was missing the word `Bearer` in the `Authorization` header. Once I fixed that, the following one-liner works, which is the cleanest way to use a custom cert file.

```
new ReactorClientHttpConnector(opt ->
                    opt.sslSupport(b -> b.trustManager(new File(properties.getCertFile()))))
```

Thanks for your time, I really appreciate it.
